src/function/cast_string_non_nested_functions.cpp

```diff
@@ -83,5 +88,94 @@ bool TryCastStringToTimestamp::tryCast<timestamp_sec_t>(const char* input, uint6
     return true;
 }
 
+static bool isDate(const std::string& str) {
+    return RE2::FullMatch(str, "\\d{4}/\\d{1,2}/\\d{1,2}") ||
+           RE2::FullMatch(str, "\\d{4}-\\d{1,2}-\\d{1,2}") ||
+           RE2::FullMatch(str, "\\d{4} \\d{1,2} \\d{1,2}") ||
+           RE2::FullMatch(str, "\\d{4}\\\\\\d{1,2}\\\\\\d{1,2}");
+}
+
+LogicalType inferMinimalTypeFromString(const std::string& str) {
+    constexpr char array_begin = common::CopyConstants::DEFAULT_CSV_LIST_BEGIN_CHAR;
+    constexpr char array_end = common::CopyConstants::DEFAULT_CSV_LIST_END_CHAR;
+    auto cpy = StringUtils::ltrim(StringUtils::rtrim(str));
+    StringUtils::toUpper(cpy);
+    if (cpy.size() == 0) {
+        return LogicalType::ANY();
+    }
+    // Boolean
+    if (cpy == "TRUE" || cpy == "FALSE") {
+        return LogicalType::BOOL();
+    }
+    // the reason we're not going to try to match to a minimal width integer
+    // is because if we're infering the type of integer from a sequence of
+    // increasing integers, we're bound to underestimate the width
+    // if we only sniff the first few elements
+    // integer
+    if (RE2::FullMatch(cpy, "(-?0)|(-?[1-9]\\d*)")) {
+        if (cpy.size() >= 1 + NumericLimits<int128_t>::digits()) {
+            return LogicalType::DOUBLE();
+        }
+        int128_t val;
+        if (!trySimpleInt128Cast(cpy.c_str(), cpy.length(), val)) {
+            return LogicalType::STRING();
+        }
+        if (val >= NumericLimits<int64_t>::minimum()) {
+            return LogicalType::INT64();
+        }
+        return LogicalType::INT128();
+    }
+    // Real value checking
+    if (RE2::FullMatch(cpy, "(\\+|-)?(0|[1-9]\\d*)?\\.(\\d*)")) {
+        if (cpy[0] == '-') {
+            cpy.erase(cpy.begin());
+        }
+        if (cpy.size() <= DECIMAL_PRECISION_LIMIT) {
+            auto decimalPoint = cpy.find('.');
+            KU_ASSERT(decimalPoint != std::string::npos);
+            return LogicalType::DECIMAL(cpy.size() - 1, cpy.size() - decimalPoint - 1);
+        } else {
+            return LogicalType::DOUBLE();
+        }
+    }
+    // date
+    if (isDate(cpy)) {
```

Comment:
What about uuid and interval?

Reply:
UUID I can support, but intervals are trickier. I could write a regex for matching intervals, but it could be quite complex, and I couldn't guarantee for it to be valid. The current casting from string to interval doesn't support any form of `tryCast`. Instead, it casts and throws an exception when it can't, so I can't use that unless I want a performance hit to sniffing.